cmd/oceantv/broadcast_states.go

```diff
@@ -34,10 +45,18 @@ const (
 )
 
 func (ctx *broadcastContext) logAndNotify(kind notify.Kind, msg string, args ...interface{}) {
-	logForBroadcast(ctx.cfg, msg, args...)
-	err := notifier.Send(context.Background(), ctx.cfg.SKey, kind, fmtForBroadcastLog(ctx.cfg, msg, args...))
+	ctx.log(msg, args...)
+	// If context has nil notifier, use global notifier
+	if ctx.notifier == nil {
+		ctx.log("broadcast context notifier is nil, setting to global notifier")
+		if notifier == nil {
+			panic("global notifier is nil")
```

Comment:
IMO, when it comes to long-running services such as this, panics should only happen in startup/setup code.

In this case, the global notifier is guaranteed to exist since it is instantiated in setup (or panics there if it is not).

Reply:
AFAIU it's idiomatic to use a panic for issues that are either programmer created (which can also categorise the following two instances), when it's impossible to continue execution or do anything about the problem (or we don't have responsibility to do anything about the problem), or when execution should not continue because the software is now dysfunctional in an unacceptable way.

The global notifier being non nil is a fundamental assumption that should hold, and we don't want to continue execution if it's nil (which can hide issues for some period unnecessarily). We also shouldn't need to create it here because that's not this codes responsibility. 

This code shouldn't make assumptions about how the global is set, or any other error checking pertinent to it. This goes back to the whole globals are dangerous philosophy; their state can change unexpectedly. You could make the argument that because it's guaranteed to exist then that is precisely the reason we should use a panic i.e. "this should definitely not happen and it's unexpected if it does, and we want to know about it".

Reply:
I'm going to resolve this conversation and merge this change. If we still have a problem with use of panics, perhaps we can create an issue and discuss there :smile: